### PR TITLE
Added an auth solution for websockets

### PR DESCRIPTION
**Pull request by sedge**
Until we figure out how to intercept the HTTP get request that is upgraded into a websocket, we will be using a token-based system to indicate that a websocket can be trusted. Anyone can open a websocket with the server with this approach, but only authenticated users will be able to get a token allowing them to use it.

### CODE REVIEW

**Comment by sedge** [comment]
@yoavgurevich Thanks for the feedback - in the future, click the `files` tab and comment directly on the file changes. 

By clicking the commit and then making comments it makes it very difficult for me to see what you're commenting on.

File: server/lib/socket-server.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 var Sync = require( './sync'),
     SyncMessage = require('./syncmessage'),
-    WebSocketServer = require('ws').Server;
+    WebSocketServer = require('ws').Server,
+    wsAuthStore = require('./wsAuthStore');
```
Comment (humphd) [inline server/lib/socket-server.js]: This file name isn't great.  Always use lowercase filenames.  Also, why switch from `WebSocket*` as a prefix to `ws` here?  I'd be consistent.

File: server/lib/socket-server.js [hunk 2]
```diff
@@ -11,43 +12,54 @@ module.exports = function( server ) {
...
-         }
-       }
-     });
-   });
+    ws.once('message', function(data, flags) {
+      // Socket data sent from a web browser
+      // is accessed through `data.data`, whereas
+      // requests sent from the NodeJS `request` module
+      // are accessed through `data`.
+      if (typeof data !== "string") {
```
Comment (humphd) [inline server/lib/socket-server.js]: This `typeof` check is odd.  What about doing `data = data.data || data` instead (i.e., just grab the `data` property if there, otherwise use the whole thing)?

File: server/lib/socket-server.js [hunk 3]
```diff
@@ -11,43 +12,54 @@ module.exports = function( server ) {
...
+
+      var syncId = data.syncId;
+      if ( !syncId ) {
+        return ws.close(1008, "SyncId required");
+      }
+
+      // Authorize user
+      var token = data.token;
+      if ( !token || !wsAuthStore.authorizeUser(token) ) {
+        return ws.close(1008, "Valid auth token required");
```
Comment (humphd) [inline server/lib/socket-server.js:37]: You re-use 1008 here (for invalid token and invalid syncID)...is that intended?

File: server/lib/socket-server.js [hunk 4]
```diff
@@ -11,43 +12,54 @@ module.exports = function( server ) {
...
+      }
+      var sync;
+
+      ws.send(JSON.stringify(new SyncMessage(SyncMessage.RESPONSE, SyncMessage.ACK)));
+
+      sync = Sync.retrieve( syncId );
+      sync.addSocket( this );
```
Comment (humphd) [inline server/lib/socket-server.js]: Probably `setSocket` unless you can actually have more then one.

File: server/lib/socket-server.js [hunk 5]
```diff
@@ -11,43 +12,54 @@ module.exports = function( server ) {
...
+
+      ws.on('message', function(data, flags) {
+        if(!flags || (flags && !flags.binary)) {
+          if (typeof data !== "string") {
```
Comment (humphd) [inline server/lib/socket-server.js]: Same comment here.  Don't use typeof if what you really mean to do is just reach in and get data.

Comment (humphd) [inline server/lib/socket-server.js]: Another comment, this code repeats above.  Why not make it a function you reuse?

File: server/lib/socket-server.js [hunk 6]
```diff
@@ -11,43 +12,54 @@ module.exports = function( server ) {
...
+            data = data.data;
+          }
+          try {
+            data = JSON.parse(data);
+            sync.messageHandler(data);
+          } catch(error) {
+            var Error = new SyncMessage(SyncMessage.RESPONSE, SyncMessage.ERROR);
```
Comment (humphd) [inline server/lib/socket-server.js]: This whole error block also seems like something you could do in a function.

File: server/lib/wsAuthStore.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+var uuid = require('node-uuid');
+
+/**
+ * Shared datastore of trusted users and their associated
```
Comment (humphd) [inline server/lib/wsAuthStore.js]: I'd spend some more time documenting what this all means.  What does "trusted" mean in our case?  You should talk about how we expect this to work with the session cookie, http auth, token, etc.  Spell it out.

File: server/lib/wsAuthStore.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * one-time use websocket auth tokens.
+ */
+var authTable = {};
+
+module.exports = {
+  addToken: function(username) {
```
Comment (humphd) [inline server/lib/wsAuthStore.js]: The naming here is odd.  You are adding a token, but taking a username.  Sounds like `generateTokenForUser` to me.

File: server/lib/wsAuthStore.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    var user = authTable[username];
+    var token = uuid.v4();
+
+    if (!user) {
+      user = authTable[username] = {
+        tokens: []
+      };
+    }
+
+    user.tokens.push(token);
```
Comment (humphd) [inline server/lib/wsAuthStore.js]: Where's the code to remove these when users disconnect?

File: server/lib/wsAuthStore.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    return token;
+  },
+  authorizeUser: function(token) {
```
Comment (humphd) [inline server/lib/wsAuthStore.js]: Extra newline above this.

Comment (humphd) [inline server/lib/wsAuthStore.js]: This function is named poorly.  You're not authorizing, you seeing if they are authorized--`isAuthorized(token)`.

File: server/lib/wsAuthStore.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    var user,
+        username,
+        i;
+
+    for (username in authTable) {
+      user = authTable[username];
+
+      for (i = 0; i < user.tokens.length; i++) {
+        if (user.tokens[i] === token) {
+          delete user.tokens[i];
```
Comment (humphd) [inline server/lib/wsAuthStore.js]: You can't remove things from an array like this.  You have to use splice.

File: server/lib/socket-server.js (hunk 2 above)
Comment (sedge) [inline server/lib/socket-server.js]: I like it! Will do.

File: server/lib/wsAuthStore.js (hunk 11 above)
Comment (humphd) [inline server/lib/wsAuthStore.js]: I'm also unclear why you're removing this at all.  Can you document the logic here?

File: server/routes/routes/index.js [hunk 12]
```diff
@@ -1,6 +1,7 @@
 var env = require('../../lib/environment'),
     version = require('../../../package.json').version,
-    FilerWebServer = require('../../lib/filer-www.js');
+    FilerWebServer = require('../../lib/filer-www.js'),
+    wsAuthStore = require('../../lib/wsAuthStore');
```
Comment (humphd) [inline server/routes/routes/index.js]: Same naming issues.

File: server/routes/routes/index.js [hunk 13]
```diff
@@ -20,5 +21,8 @@ module.exports = {
       http: "okay",
       version: version
     });
+  },
+  generateToken: function( req, res ) {
+    res.json(200, wsAuthStore.addToken(req.params.username));
```
Comment (humphd) [inline server/routes/routes/index.js]: Notice how your route is `generateToken` and you have a method called `addToken`.  This sort of thing leads to confusion.  Stick with `generate`

File: server/lib/socket-server.js (hunk 3 above)
Comment (sedge) [inline server/lib/socket-server.js:37]: Yes. Both indicate someone breaking the protocol we've set up by sending or omitting invalid data.

Honestly, the error codes are very abstract and not that useful. This was the best I could find, though Yoav also suggested 1011 for unexpected exception.

File: tests/lib/util.js [hunk 14]
```diff
@@ -167,6 +167,21 @@ function getConnectionID(options, callback){
   });
 }
 
+function getWsToken(options, callback){
```
Comment (humphd) [inline tests/lib/util.js]: Why do you start calling it `WsToken` when everywhere else you have used just `token`.  Consistency.

File: server/lib/socket-server.js (hunk 4 above)
Comment (sedge) [inline server/lib/socket-server.js]: :+1:

File: tests/unit/routes.js [hunk 15]
```diff
@@ -153,3 +153,31 @@ describe('/healthcheck', function () {
     });
   });
 });
+
+describe('/sync/token', function() {
+  it('should return a token when a user is authenticated', function(done) {
+    util.authenticate({ username: util.username() }, function(err, authResult) {
```
Comment (humphd) [inline tests/unit/routes.js]: Unless you need to use the username elsewhere in the test, it's optional to pass it here (one will get created automatically).

File: tests/unit/routes.js [hunk 16]
```diff
@@ -153,3 +153,31 @@ describe('/healthcheck', function () {
...
+      request({
+        url: util.serverURL + '/api/sync',
+        jar: authResult.jar,
+        json: true
+      }, function(err, response, body) {
+        expect(err).not.to.exist;
+        expect(response.statusCode).to.equal(200);
+        expect(body).to.be.a('string');
```
Comment (humphd) [inline tests/unit/routes.js]: This seems like a weak test.  Can you do more than just expect a string?

File: tests/unit/routes.js [hunk 17]
```diff
@@ -153,3 +153,31 @@ describe('/healthcheck', function () {
...
+
+        done();
+      });
+    });
+  });
+  it('should reject a request with a 401 if the user is not authenticated', function(done) {
```
Comment (humphd) [inline tests/unit/routes.js]: Insert newline above this.

File: server/lib/websocket-auth.js [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+ *     and can no longer be used.
+ *
+ * authorizeToken(token)
+ *   - If the token passed exists for a particular user, the token
+ *     is deleted from the datastore, returning true.
+ *
+ * purgeSession(username, sessionId)
+ *   - Deletes all tokens for a particular user session, intended to be
+ *     used on webmakerauth signout.
+
```
Comment (Plumpernickel) [inline server/lib/websocket-auth.js]: Not sure how important this is, but you missed an `*` on this line...

Comment (sedge) [inline server/lib/websocket-auth.js]: Good catch

File: server/lib/socket-server.js [hunk 19]
```diff
@@ -11,43 +12,47 @@ module.exports = function( server ) {
...
+      }
+
+      // Authorize user
+      var token = data.token;
+      if ( !token || !websocketAuth.authorizeToken(token) ) {
+        return ws.close(1008, "Valid auth token required");
+      }
+      var sync;
+
+      ws.send(JSON.stringify(new SyncMessage(SyncMessage.RESPONSE, SyncMessage.ACK)));
```
Comment (humphd) [inline server/lib/socket-server.js]: This is a lot of code you have to write over and over.  Consider a convenience:

``` javascript
// Create a new Response/ACK SyncMessage as a JSON string
ws.send(SyncMessage.Response(SyncMessage.ACK));

// or even...provide pre-created versions
ws.send(SyncMessage.Response.ACK);
```

Know what I mean?

File: server/lib/socket-server.js [hunk 20]
```diff
@@ -11,43 +12,47 @@ module.exports = function( server ) {
...
+      sync = Sync.retrieve( syncId );
+      sync.setSocket( ws );
+
+      ws.on('message', function(data, flags) {
+        if(!flags || (flags && !flags.binary)) {
+          try {
+            data = JSON.parse(data);
+            sync.messageHandler(data);
+          } catch(error) {
+            this.send(JSON.stringify(SyncMessage.generateError(error)));
```
Comment (humphd) [inline server/lib/socket-server.js]: Prefer `ws` to `this`.

File: server/lib/websocket-auth.js [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+ * authorizeToken(token)
+ *   - If the token passed exists for a particular user, the token
+ *     is deleted from the datastore, returning true.
+ *
+ * purgeSession(username, sessionId)
+ *   - Deletes all tokens for a particular user session, intended to be
+ *     used on webmakerauth signout.
+ */
+var authTable = {};
+var timeout = env.get("TOKEN_TIMEOUT") || 60000; // Default to 60 sec
```
Comment (humphd) [inline server/lib/websocket-auth.js]: When you have a timeout, it's good to specify the units in the name.  Since this is milliseconds, use TOKEN_TIMEOUT_MS

File: server/lib/websocket-auth.js [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+function createSessionTracker() {
+  var sessionId = uuid.v4();
+
+  authTable[sessionId] = [];
+
+  return sessionId;
```
Comment (humphd) [inline server/lib/websocket-auth.js:35]: Drop all the newlines in this function.

File: server/lib/websocket-auth.js [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+function generateTokenForSession(sessionId) {
+  var sessionData = authTable[sessionId];
```
Comment (humphd) [inline server/lib/websocket-auth.js:39]: What if `authTable[sessionId]` doesn't exist?  does that matter to you?  you don't do any check.

File: server/lib/websocket-auth.js (hunk 21 above)
Comment (humphd) [inline server/lib/websocket-auth.js]: Don't name this variable `timeout`, name it the same as the env variable.

File: server/lib/websocket-auth.js [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+    }
+  }
+  return false;
+}
+
+function purgeSession(sessionId) {
+  var sessionData = authTable[sessionId];
+
+  if (sessionData) {
+    delete sessionData;
```
Comment (humphd) [inline server/lib/websocket-auth.js]: You can't do this.  Delete is meant to be used on an object, not a variable in this scope.  You probably want to remove it from authTable.

File: server/routes/middleware/index.js [hunk 25]
```diff
@@ -13,7 +14,13 @@ module.exports = {
       return next( generateError( 401, "Webmaker Authentication Required." ) );
     }
 
+    var sessionId = req.session.sessionId;
```
Comment (humphd) [inline server/routes/middleware/index.js]: These lines feel more complicated than they need to be.  Why not just check if `req.session.sessionId` is there, if not, generate it, then assign to `sessionId`?

File: tests/lib/util.js [hunk 26]
```diff
@@ -167,6 +167,21 @@ function getConnectionID(options, callback){
   });
 }
 
+function getWebsocketToken(options, callback){
+  if(!(options && options.jar)) {
+    throw('Expected options.jar');
```
Comment (humphd) [inline tests/lib/util.js:173]: Why throw instead of send the error back on the callback?

File: tests/lib/util.js [hunk 27]
```diff
@@ -167,6 +167,21 @@ function getConnectionID(options, callback){
...
+  }
+
+  request({
+    url: serverURL + '/api/sync',
+    jar: options.jar,
+    json: true
+  }, function(err, response, body) {
+    options.token = body;
+    callback(err, options);
```
Comment (humphd) [inline tests/lib/util.js]: This is a dangerous pattern.  You should first deal with the `err` case, call the callback and return, then deal with the non-err case(s).

File: tests/unit/routes.js [hunk 28]
```diff
@@ -13,142 +13,172 @@ beforeEach(function(){
...
-        result2.done();
+        result1.syncId = uuid.v4();
+        util.syncRouteConnect(result1, function(err, result2){
+          expect(err).not.to.exist;
+          expect(result2.statusCode).to.equal(400);
+          result2.done();
+        });
       });
     });
+    // Can we test against sync.start's fatal mkdir error/500 statusCode?
```
Comment (humphd) [inline tests/unit/routes.js]: Why is this here?

File: tests/unit/routes.js [hunk 29]
```diff
@@ -13,142 +13,172 @@ beforeEach(function(){
...
+    it('should return a 400 when path is missing from request.body', function (done){
+      util.authenticatedConnection({done: done}, function(err, result1) {
+        expect(err).not.to.exist;
+          util.syncRouteConnect(result1, function(err, result2) {
             expect(err).not.to.exist;
-            expect(result3.statusCode).to.equal(400);
-            result3.done();
-          });
+            var localData = data;
+            localData.json.path = undefined;
```
Comment (humphd) [inline tests/unit/routes.js]: You typically only use `undefined` for comparisons, not to assign to a variable.  You use `null` instead.

File: tests/unit/routes.js [hunk 30]
```diff
@@ -13,142 +13,172 @@ beforeEach(function(){
...
+    it('should return a 400 when the srcList is missing from request.body', function (done){
+      util.authenticatedConnection({done: done}, function(err, result1) {
+        expect(err).not.to.exist;
+          util.syncRouteConnect(result1, function(err, result2) {
             expect(err).not.to.exist;
-            expect(result3.statusCode).to.equal(400);
-            result3.done();
-          });
+            var localData = data;
+            localData.json.srcList = undefined;
```
Comment (humphd) [inline tests/unit/routes.js]: If you really want this to be `undefined`, you would usually just delete it, `delete localData.json.srcList;`

File: tests/unit/routes.js [hunk 31]
```diff
@@ -13,142 +13,172 @@ beforeEach(function(){
...
+          json: true
+        }, function(err, res, body) {
           expect(err).not.to.exist;
-          util.csRouteConnect(result3, function(err, result4) {
-            expect(err).not.to.exist;
-            expect(result4.body.checksums.length).to.be.above(0);
-            expect(result4.statusCode).to.equal(200);
-            result4.done();
-          });
+          expect(res.body.http).to.deep.equal('okay');
```
Comment (humphd) [inline tests/unit/routes.js]: Why deep equal on a string?

File: tests/lib/util.js (hunk 26 above)
Comment (sedge) [inline tests/lib/util.js:173]: Good question. I was following your example from `getConnectionID(opts, cb)`, but I can't say I thought the decision through.

I'd say that it's outside of the scope of the test itself - in other words, there's no way the tests using this can run properly without `options.jar` being passed, so it interrupting execution is probably preferable no? Either that or we can safely assume it'll be passed in.

File: server/lib/websocket-auth.js (hunk 23 above)
Comment (sedge) [inline server/lib/websocket-auth.js:39]: I'm fairly sure it can be safely assumed because of the checks that happen in the middleware chain before this function is called.

**Comment by sedge** [comment]
@humphd says: r=me
